src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+        }
+        CreateFunctionDataType::TEXT
+        | CreateFunctionDataType::CHAR
+        | CreateFunctionDataType::VARCHAR => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::StringArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to string array",
+                col_ix
+            ))),
+        },
+        CreateFunctionDataType::BOOLEAN => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::BooleanArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to boolean array",
+                col_ix
+            ))),
+        },
+        // serialize decimal as string
+        CreateFunctionDataType::DECIMAL {
+            precision: _,
+            scale: _,
+        } => get_arrow_value::<arrow::datatypes::Decimal128Type>(args, row_ix, col_ix)
+            .map(|val| Value::from(val.to_string())),
+        // dates are represented as i32 integer internally, serialize them as such.
+        CreateFunctionDataType::DATE => {
+            get_arrow_value::<arrow::datatypes::Date32Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_string()))
+        }
+        // timestamps are represented as i64 integers internally, serialize them as such.
+        CreateFunctionDataType::TIMESTAMP => get_arrow_value::<
+            arrow::datatypes::TimestampNanosecondType,
+        >(args, row_ix, col_ix)
+        .map(|val| Value::from(val.to_string())),
+    }
+}
+
+fn decode_udf_result_primitive_array<T>(
+    encoded_results: &Vec<Value>,
+    decoder: &dyn Fn(&Value) -> Result<T::Native>,
+) -> Result<PrimitiveArray<T>>
+where
+    T: ArrowPrimitiveType,
+{
+    let mut decoded_results = Vec::with_capacity(encoded_results.len());
+    for i in encoded_results {
+        decoded_results.push(Some(decoder(i)?));
+    }
+    Ok(decoded_results.iter().collect::<PrimitiveArray<T>>())
+}
+
+fn messagepack_decode_results(
+    return_type: &CreateFunctionDataType,
+    encoded_results: &Vec<Value>,
+) -> Result<ArrayRef> {
+    match return_type {
+        CreateFunctionDataType::SMALLINT => decode_udf_result_primitive_array::<
+            arrow::datatypes::Int16Type,
+        >(encoded_results, &|v| match v
+            .as_i64()
+            .ok_or(DataFusionError::Internal(format!(
+                "Expected to find i64 value, but received {:?} instead",
+                v
+            ))) {
+            Err(e) => Err(e),
+            Ok(v_i64) => i16::try_from(v_i64).map_err(|e| {
+                DataFusionError::Internal(format!("Error converting i64 to i16: {:?}", e))
+            }),
+        })
```

Comment:
You don't need this `match` here either, you can do it more cleanly as:

```suggestion
        >(encoded_results, &|v| v
            .as_i64()
            .ok_or(DataFusionError::Internal(format!(
                "Expected to find i64 value, but received {:?} instead",
                v
            ))).and_then(|v_i64| i16::try_from(v_i64).map_err(|e| {
                DataFusionError::Internal(format!("Error converting i64 to i16: {:?}", e))
            })
```

Reply:
Definitely looks nicer, thanks!